internal-messaging/message/cc/DataMessage.cc

```diff
@@ -15,44 +15,64 @@ DataMessage::DataMessage(unsigned int sender, unsigned int recipient, long time,
 :Message(sender, recipient, time, 100, contents)
 {}
 
+DataMessage::DataMessage(string flat){
+	char cstr[flat.size() + 1];
+	std::strcpy(cstr, flat.c_str());	// or pass &s[0]
+	BuildFromCharacters(cstr);
+}
+
 DataMessage::DataMessage(char* flat)
 {
-    std::cout << "Creating Data Message" << std::endl;
-    char integer_string[32];
-    char hex_string[32];
-    memset(integer_string, 0, 32);
-    memset(hex_string, 0, 32);
+	BuildFromCharacters(flat);
+}
 
-    // Find sneder, recipient, time and flag
-    int i = BuildHeader(flat, 0);
-    i++;
-    // Skip over flag
-    while(flat[i] != '|'){
-        i++;
-    }
-    i++;
-    // Find Requests
-    while(flat[i] != '|'){
-        // Get next request
-        while(flat[i] != '-' && flat[i] != '|'){
-            sprintf(integer_string, "%c", flat[i]);
-            strcat(hex_string, integer_string); 
-            i++;
-        }
-        try {
-            int request = std::stoi(hex_string,nullptr,16);
-            requests.push_back(request);
-        } catch (std::exception const &e) {}
-        memset(integer_string, 0, 32);
-        memset(hex_string, 0, 32);
-        if(flat[i] != '|'){
-            i++;
-        };
-    }
+/*
+ * Takes a character array, extracts data, and assigns it to class variables
+ * This function assumes that flat is the "flattened" version of a pre-existing DataMessage
+ * TODO for andrew: give this a better name
+ */
+int DataMessage::BuildFromCharacters(char * flat){
+	 std::cout << "Creating Data Message" << std::endl;
+	    char integer_string[32];
+	    char hex_string[32];
+	    memset(integer_string, 0, 32);
+	    memset(hex_string, 0, 32);
+	    // Find sender, recipient, time and flag
+	    int i = BuildHeader(flat, 0);
+	    i++;
+	    //TODO Spencer, should the flag be parsed in parent class Message? - Andrew
+	    // Skip over flag
+	    while(flat[i] != '|'){
+	        i++;
+	    }
+	    //i++;
+
+	    //TODO Find more robust solution for when no requests exist
+	    // Find Requests
+	    while(flat[i] != '|' && flat[i-1] != '|'){
+	        // Get next request
+	    	//TODO fix bug where infinite loop is reached due to no requests in the message
+	        while(flat[i] != '-' && flat[i] != '|'){
```

Comment:
This seems oddly redundant to have two `while` loops with the same condition (`flat[i] != '|'` & `flat[i-1] != '|'). There should be a comment here explaining what `|` and `-` are in this loop. Maybe instead of passing a raw char pointer, a c++ string should be used instead which holds the size of the string. This would transform this loop into being something like `while(flat[i] != `i` && i < flat.size())`

Reply:
@s-axford ?